vsintegration/src/FSharp.Compiler.Server/Ipc/IpcServer.fs

```diff
@@ -0,0 +1,49 @@
+﻿namespace Microsoft.FSharp.Compiler.Server
+
+open System
+open System.IO
+open System.Diagnostics
+open System.Threading
+open System.Threading.Tasks
+open System.IO.Pipes
+
+open Newtonsoft.Json
+
+type IpcMessageServer<'Receive, 'Response>(name, f : 'Receive -> Async<'Response>) =
+
+    let buffer = Array.zeroCreate<char> Constants.IpcBufferSize
+
+    member this.Run() =
+        use fcs = new NamedPipeServerStream(name, PipeDirection.InOut, 1, PipeTransmissionMode.Message, PipeOptions.None)
+
+        printfn "[FSharp Compiler Server] - Waiting for connection"
+
+        fcs.WaitForConnection()
+
+        printfn "[FSharp Compiler Server] - Client connected"
+
+        let writer = new StreamWriter(fcs, AutoFlush = true)
+        let reader = new StreamReader(fcs)
+
+        try
+            while true do
+                if not fcs.IsConnected then
+                    failwith "Client disconnected"
+
+                let count = reader.Read(buffer, 0, buffer.Length)
```

Comment:
Better off using ReadLine/WriteLine.

It's possible that:

1. The buffer is too small to hold the entire message (and so you end up reading a partial message)
2. There are multiple messages in the buffer (and so the JSON deserialisation fails)

You can tell Newtonsoft.Json to serialise a JSON object on a single-line by using Formatting.None.